Mathlib/Logic/Relation.lean

```diff
@@ -0,0 +1,638 @@
+  ⟨trans_left⟩
+
+@[trans]
+theorem trans (hab : TransGen r a b) (hbc : TransGen r b c) : TransGen r a c :=
+  trans_left hab hbc.to_reflTransGen
+
+instance : Trans (TransGen r) (TransGen r) (TransGen r) :=
+  ⟨trans⟩
+
+theorem head' (hab : r a b) (hbc : ReflTransGen r b c) : TransGen r a c :=
+  trans_left (single hab) hbc
+
+theorem tail' (hab : ReflTransGen r a b) (hbc : r b c) : TransGen r a c := by
+  induction hab generalizing c
+  case refl => exact single hbc
+  case tail _ _ _ hdb IH => exact tail (IH hdb) hbc
+
+theorem head (hab : r a b) (hbc : TransGen r b c) : TransGen r a c :=
+  head' hab hbc.to_reflTransGen
+
+@[elab_as_elim]
+theorem head_induction_on {P : ∀ a : α, TransGen r a b → Prop} {a : α} (h : TransGen r a b)
+    (base : ∀ {a} (h : r a b), P a (single h))
+    (ih : ∀ {a c} (h' : r a c) (h : TransGen r c b), P c h → P a (h.head h')) : P a h := by
+  induction h
+  case single a h => exact base h
+  case tail b c _ hbc h_ih =>
+  -- Lean 3 could figure out the motive and `apply
+  refine @h_ih (λ {a : α} (hab : @TransGen α r a b) => P a (TransGen.tail hab hbc)) ?_ ?_;
+  exact fun h => ih h (single hbc) (base hbc)
+  exact fun hab hbc => ih hab _
+
+@[elab_as_elim]
+theorem trans_induction_on {P : ∀ {a b : α}, TransGen r a b → Prop} {a b : α} (h : TransGen r a b)
+    (base : ∀ {a b} (h : r a b), P (single h))
+    (ih : ∀ {a b c} (h₁ : TransGen r a b) (h₂ : TransGen r b c), P h₁ → P h₂ → P (h₁.trans h₂)) :
+    P h := by
+  induction h with
+  | single h => exact base h
+  | tail hab hbc h_ih => exact ih hab (single hbc) h_ih (base hbc)
+
+theorem trans_right (hab : ReflTransGen r a b) (hbc : TransGen r b c) : TransGen r a c := by
+  induction hbc
+  case single c hbc => exact tail' hab hbc
+  case tail c d _ hcd hac => exact hac.tail hcd
+
+instance : Trans (ReflTransGen r) (TransGen r) (TransGen r) :=
+  ⟨trans_right⟩
+
+theorem tail'_iff : TransGen r a c ↔ ∃ b, ReflTransGen r a b ∧ r b c := by
+  refine' ⟨fun h => _, fun ⟨b, hab, hbc⟩ => tail' hab hbc⟩
+  cases' h with _ hac b _ hab hbc
+  · exact ⟨_, by rfl, hac⟩
+
+  · exact ⟨_, hab.to_reflTransGen, hbc⟩
+
+
+theorem head'_iff : TransGen r a c ↔ ∃ b, r a b ∧ ReflTransGen r b c := by
+  refine' ⟨fun h => _, fun ⟨b, hab, hbc⟩ => head' hab hbc⟩
+  induction h
+  case single c hac => exact ⟨_, hac, by rfl⟩
+  case tail b c _ hbc IH =>
+  rcases IH with ⟨d, had, hdb⟩
+  exact ⟨_, had, hdb.tail hbc⟩
+
+end TransGen
+
+theorem _root_.Acc.TransGen (h : Acc r a) : Acc (TransGen r) a := by
+  induction' h with x _ H
+  refine' Acc.intro x fun y hy => _
+  cases' hy with _ hyx z _ hyz hzx
+  exacts[H y hyx, (H z hzx).inv hyz]
+
+theorem _root_.acc_transGen_iff : Acc (TransGen r) a ↔ Acc r a :=
+  ⟨Subrelation.accessible TransGen.single, Acc.TransGen⟩
+#align acc_trans_gen_iff acc_transGen_iff
+
+theorem _root_.well_founded.transGen (h : WellFounded r) : WellFounded (TransGen r) :=
+  ⟨fun a => (h.apply a).TransGen⟩
+
+#align well_founded.trans_gen well_founded.transGen
```

Comment:
```suggestion
theorem _root_.WellFounded.transGen (h : WellFounded r) : WellFounded (TransGen r) :=
  ⟨fun a => (h.apply a).TransGen⟩
#align well_founded.trans_gen WellFounded.transGen
```

Reply:
Thanks! I know you told me this today on Zulip as well, but I've not looked at this PR or indeed at github for the last week because of work commitments.